kernel/uid_gid.c

```diff
@@ -780,3 +837,139 @@ long myst_syscall_setgroups(size_t size, const gid_t* list)
 
     return 0;
 }
+
+int check_thread_group_membership(gid_t group)
+{
+    myst_thread_t* thread = myst_thread_self();
+    if (group == thread->egid)
+        return 1;
+
+    for (unsigned int i = 0; i < thread->num_supgid; i++)
+    {
+        if (group == thread->supgid[i])
+            return 1;
+    }
+
+    return 0;
+}
+
+long myst_syscall_chown(const char* pathname, uid_t owner, gid_t group)
+{
+    long ret = 0;
+    myst_fs_t* fs;
+    myst_thread_t* thread = myst_thread_self();
+    struct locals
+    {
+        char suffix[PATH_MAX];
+        struct stat statbuf;
+    }* locals = NULL;
+
+    if (!pathname)
+        return -EINVAL;
+
+    if (((owner != (uid_t)-1) &&
+         (myst_valid_uid_against_passwd_file(owner) <= 0)) ||
+        ((group != (gid_t)-1) &&
+         (myst_valid_gid_against_group_file(group) <= 0)))
+    {
+        ret = -EINVAL;
+    }
+
+    if (!(locals = malloc(sizeof(struct locals))))
+        ERAISE(-ENOMEM);
+
+    ECHECK(myst_mount_resolve(pathname, locals->suffix, &fs));
+
+    /* if thread's euid is root (TODO: or has CAP_CHOWN capability) */
+    if (thread->euid == 0)
+    {
+        ECHECK((*fs->fs_chown)(fs, locals->suffix, owner, group));
+    }
+    /* non-privileged thread case */
+    else
+    {
+        /* file should be owned by the thread */
+        ECHECK((*fs->fs_stat)(fs, locals->suffix, &locals->statbuf));
+        if (locals->statbuf.st_uid != thread->euid)
+            ERAISE(-EPERM);
+
+        /* owner should be -1 or user ID of file */
+        if (!(owner == -1u || owner == locals->statbuf.st_uid))
+            ERAISE(-EPERM);
+
+        /* group should either be thread's egid or one of the supplementary gids
+         */
+        if (!check_thread_group_membership(group))
+            ERAISE(-EPERM);
+
+        ECHECK(fs->fs_chown(fs, locals->suffix, owner, group));
+    }
+
+done:
+
+    if (locals)
+        free(locals);
+
+    return ret;
+}
+
+long myst_syscall_fchown(int fd, uid_t owner, gid_t group)
+{
+    long ret = 0;
+    myst_file_t* file = NULL;
+    myst_fs_t* fs = NULL;
+    myst_fdtable_t* fdtable = myst_fdtable_current();
+    myst_thread_t* thread = myst_thread_self();
+    struct locals
+    {
+        char suffix[PATH_MAX];
+        struct stat statbuf;
+    }* locals = NULL;
+
+    if (fd < 0)
+        return -EINVAL;
+
+    if (((owner != (uid_t)-1) &&
+         (myst_valid_uid_against_passwd_file(owner) <= 0)) ||
+        ((group != (gid_t)-1) &&
+         (myst_valid_gid_against_group_file(group) <= 0)))
+    {
+        ret = -EINVAL;
+    }
+
+    if (!(locals = malloc(sizeof(struct locals))))
+        ERAISE(-ENOMEM);
+
+    ECHECK(myst_fdtable_get_file(fdtable, fd, &fs, &file));
+
+    /* if thread's euid is root (TODO: or has CAP_CHOWN capability) */
+    if (thread->euid == 0)
```

Comment:
Can we consolidate `myst_syscall_fchown` and `myst_syscall_chown` to avoid the duplicate code?

Reply:
Have moved some of the common code into a helper function.